samples: dfu: Add A/B sample (#24443)
nrfconnect/sdk-nrf: samples/dfu/ab/README.rst

You can update the sample using SMP protocol with UART or Bluetooth® Low Energy.

Requirements
************

The sample supports the following development kit:

.. table-from-sample-yaml::

You need the nRF Device Manager app for SUIT update over Bluetooth Low Energy:

* `nRF Device Manager mobile app for Android`_
* `nRF Device Manager mobile app for iOS`_

Overview
********

This sample demonstrates firmware update using "A/B" method.
This method allows to have two copies of the application in the NVM memory.
It is possible to switch between these copies, without the need to perform swap what significantly reduces time of device's unavailability during the update.
Switch between images can be triggered by application or for example HW button.

This sample application implements a Simple Management Protocol (SMP) server.
SMP is a basic transfer encoding for use with the MCUmgr management protocol.
For more information about MCUmgr and SMP, please see :ref:`device_mgmt`.

This sample application supports the following MCUmgr transports by default:

    * Bluetooth
    * Serial (UART)

Configuration
*************

> The :kconfig:option:`CONFIG_N_BLINKS` Kconfig option sets the number of times the LED corresponding to the currently active slot blinks (LED0 for slot A, LED1 for slot B).

Reviewer: Edits are done to follow the template https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/templates/sample_README.html#configuration_options.
```suggestion
Configuration options
=====================

Check and configure the following configuration option for the sample:

.. _CONFIG_N_BLINKS:

CONFIG_N_BLINKS - The number of blinks.
   This configuration option sets the number of times the LED corresponding to the currently active slot blinks (LED0 for slot A, LED1 for slot B).
```

> The default value of the option is set to ``1``, causing a single blink to indicate *Version 1*.
> You can increment this value to represent an update, such as set it to ``2`` to indicate *Version 2*.

Reviewer: ```suggestion
   The default value of the option is set to ``1``, causing a single blink to indicate *Version 1*.
   You can increment this value to represent an update, such as set it to ``2`` to indicate *Version 2*.
```

> Use :kconfig:option:`CONFIG_APP_HEALTH_CHECK_FAILURE` Kconfig option to emulate a "broken" application that does not pass the self-test.

Reviewer: ```suggestion
.. _CONFIG_APP_HEALTH_CHECK_FAILURE:

CONFIG_APP_HEALTH_CHECK_FAILURE - Enables emulation of a broken application that fails the self-test 
```

> To ensure the updated build is preferred after a DFU, set the :kconfig:option:`CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION` option to a higher version number then currently running on the device.

Reviewer: ```suggestion
:kconfig:option:`CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION` - This configuration option sets the version to pass to imgtool when signing.
To ensure the updated build is preferred after a DFU, set this option to a higher version than the version currently running on the device.
```

Failing to do so may result in the image not being selected on the next reboot, depending on the slot used for the update.

User interface
**************

.. tabs::

   .. group-tab:: nRF54 DKs

      LED 0:
         This LED indicates that the application is running from slot A.
         This LED is controlled as "active low", meaning it will turn on once the application is booted and blinks (turns off) in short intervals.
         The number of short blinks is configurable using the :kconfig:option:`CONFIG_N_BLINKS` Kconfig option.
         It will be off if the application is running from slot B.

      LED 1:
         This LED indicates that the application is running from slot B.

>          This LED is controlled as "active low", meaning it will turn on once the application is booted and blinks (turns off) in short intervals.

Reviewer: ```suggestion
         It is controlled as "active low", meaning it will turn on once the application is booted and blinks (turns off) in short intervals.
```

         The number of short blinks is configurable using the :kconfig:option:`CONFIG_N_BLINKS` Kconfig option.
         It will be off if the application is running from slot A.

      Button 0:
         By pressing this button, the non-active slot will be selected as the preferred slot on the next reboot.
         This preference applies only to the next boot and will be cleared after the subsequent reset.

Building and running
********************

> .. |sample path| replace:: :file:`samples/dfu/dfu_target`

Reviewer: This path needs to be changed.

.. include:: /includes/build_and_run.txt

Testing
=======

> To perform DFU using the `nRF Connect Device Manager`_ mobile app, complete the following steps:

Reviewer: Please check if this edit is appropriate.
```suggestion
|test_sample|

1. |connect_terminal_specific|
2. To perform DFU using the `nRF Connect Device Manager`_ mobile app, complete the following steps:
```

.. tabs::

   .. tab:: MCUboot

      .. include:: /app_dev/device_guides/nrf52/fota_update.rst
         :start-after: fota_upgrades_over_ble_nrfcdm_common_dfu_steps_start
         :end-before: fota_upgrades_over_ble_nrfcdm_common_dfu_steps_end

Dependencies
************

> This sample uses the following |NCS| libraries:

Reviewer: ```suggestion
This sample uses the following |NCS| library:
```
